File: algorithms/152.maximum-product-subarray.cpp
class Solution {
public:
    // DP approach : O(n)
    int maxProduct(vector<int>& nums) {
        // edge case
        if (nums.size() == 0)
            return 0;
            
        int g_max = nums[0], DP_max = nums[0], DP_min = nums[0];
        for (int i=1; i < nums.size(); i++){
            int _max = max(max(DP_max*nums[i], DP_min*nums[i]), nums[i]);
            int _min = min(min(DP_max*nums[i], DP_min*nums[i]), nums[i]);
            DP_max = _max;
            DP_min = _min;
            g_max = max(g_max, DP_max);
        }
        return g_max;
    }
};

File: algorithms/152.maximum-product-subarray.java
public class Solution {
	public int maxProduct(int[] nums) {
		// Edge case
		if (nums == null || nums.length == 0) {
			return 0;
		}
		
		double maxDP = nums[0];
		double minDP = nums[0];
		double maxGlobal = nums[0];
		
		// Iterate through the array starting from the second element
		for (int i = 1; i < nums.length; i++) {
			// Store the current maximum ending here before updating
			double tempMax = maxDP;
			
			// Update the maximum and minimum ending here
			double num = nums[i];
			maxDP = Math.max(Math.max(num, maxDP * num), minDP * num);
			minDP = Math.min(Math.min(num, tempMax * num), minDP * num);
			
			// Update the maximum product found so far
			maxGlobal = Math.max(maxGlobal, maxDP);
		}
		
		return (int) maxGlobal;
	}
}
